File: go/mechanisms/evm/batch-settlement/client/permit2.go
package client

import (
	"context"
	"fmt"
	"math/big"
	"time"

	"github.com/x402-foundation/x402/go/mechanisms/evm"
	batchsettlement "github.com/x402-foundation/x402/go/mechanisms/evm/batch-settlement"
	"github.com/x402-foundation/x402/go/types"
)

// CreateBatchedPermit2DepositPayload builds a deposit + voucher payload that
// funds the channel via the universal Permit2 contract using a channel-bound
// `PermitWitnessTransferFrom` authorization. The witness binds the transfer to
// the derived channelId so the Permit2DepositCollector can verify which
// channel the funds belong to.
func CreateBatchedPermit2DepositPayload(
	ctx context.Context,
	signer evm.ClientEvmSigner,
	requirements types.PaymentRequirements,
	channelConfig batchsettlement.ChannelConfig,
	depositAmount string,
	maxClaimableAmount string,
	voucherSigner evm.ClientEvmSigner,
) (types.PaymentPayload, error) {
	networkStr := string(requirements.Network)

	chainId, err := evm.GetEvmChainId(networkStr)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to get chain ID: %w", err)
	}

	depositBig, ok := new(big.Int).SetString(depositAmount, 10)
	if !ok {
		return types.PaymentPayload{}, fmt.Errorf("invalid deposit amount: %s", depositAmount)
	}

	nonce, err := evm.CreatePermit2Nonce()
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to create permit2 nonce: %w", err)
	}
	nonceBig, ok := new(big.Int).SetString(nonce, 10)
	if !ok {
		return types.PaymentPayload{}, fmt.Errorf("invalid permit2 nonce: %s", nonce)
	}

	deadline := fmt.Sprintf("%d", time.Now().Unix()+int64(requirements.MaxTimeoutSeconds))
	deadlineBig, ok := new(big.Int).SetString(deadline, 10)
	if !ok {
		return types.PaymentPayload{}, fmt.Errorf("invalid permit2 deadline: %s", deadline)
	}

	channelId, err := batchsettlement.ComputeChannelId(channelConfig, chainId)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to compute channel ID: %w", err)
	}
	channelIdBytes, err := evm.HexToBytes(channelId)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to parse channel id: %w", err)
	}

	tokenAddress := evm.NormalizeAddress(requirements.Asset)
	spenderAddress := evm.NormalizeAddress(batchsettlement.Permit2DepositCollectorAddress)

	domain := evm.TypedDataDomain{
		Name:              batchsettlement.Permit2DomainName,
		ChainID:           chainId,
		VerifyingContract: batchsettlement.Permit2Address,
	}

	// EIP-712 type definitions: include EIP712Domain alongside the witness types
	// so the signer can hash the domain separator the same way Permit2 does.
	typedDataTypes := map[string][]evm.TypedDataField{
		"EIP712Domain": {
			{Name: "name", Type: "string"},
			{Name: "chainId", Type: "uint256"},
			{Name: "verifyingContract", Type: "address"},
		},
		"PermitWitnessTransferFrom": batchsettlement.BatchPermit2WitnessTypes["PermitWitnessTransferFrom"],
		"TokenPermissions":          batchsettlement.BatchPermit2WitnessTypes["TokenPermissions"],
		"DepositWitness":            batchsettlement.BatchPermit2WitnessTypes["DepositWitness"],
	}

	message := map[string]interface{}{
		"permitted": map[string]interface{}{
			"token":  tokenAddress,
			"amount": depositBig,
		},
		"spender":  spenderAddress,
		"nonce":    nonceBig,
		"deadline": deadlineBig,
		"witness": map[string]interface{}{
			"channelId": channelIdBytes,
		},
	}

	signature, err := signer.SignTypedData(ctx, domain, typedDataTypes, "PermitWitnessTransferFrom", message)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to sign permit2 authorization: %w", err)
	}

	actualVoucherSigner := signer
	if voucherSigner != nil {
		actualVoucherSigner = voucherSigner
	}
	voucher, err := SignVoucher(ctx, actualVoucherSigner, channelId, maxClaimableAmount, networkStr)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to sign voucher: %w", err)
	}

	depositPayload := &batchsettlement.BatchSettlementDepositPayload{
		Type:          "deposit",
		ChannelConfig: channelConfig,
		Voucher:       *voucher,
		Deposit: batchsettlement.BatchSettlementDepositData{
			Amount: depositAmount,
			Authorization: batchsettlement.BatchSettlementDepositAuthorization{
				Permit2Authorization: &batchsettlement.BatchSettlementPermit2Authorization{
					From: signer.Address(),
					Permitted: batchsettlement.BatchSettlementPermit2TokenPermissions{
						Token:  tokenAddress,
						Amount: depositAmount,
					},
					Spender:  spenderAddress,
					Nonce:    nonce,
					Deadline: deadline,
					Witness: batchsettlement.BatchSettlementPermit2Witness{
						ChannelId: channelId,
					},
					Signature: evm.BytesToHex(signature),
				},
			},
		},
	}

	return types.PaymentPayload{
		X402Version: 2,
		Payload:     depositPayload.ToMap(),
	}, nil
}

File: typescript/packages/mechanisms/evm/src/batch-settlement/client/permit2.ts
import { PaymentRequirements, PaymentPayloadResult } from "@x402/core/types";
import { getAddress } from "viem";
import { PERMIT2_ADDRESS } from "../../constants";
import { ClientEvmSigner } from "../../signer";
import { createPermit2Nonce, getEvmChainId } from "../../utils";
import { PERMIT2_DEPOSIT_COLLECTOR_ADDRESS, batchPermit2WitnessTypes } from "../constants";
import { ChannelConfig, BatchSettlementDepositPayload } from "../types";
import { computeChannelId } from "../utils";
import { signVoucher } from "./voucher";

/**
 * Builds a batch deposit payload using a channel-bound Permit2 witness transfer.
 *
 * @param signer - Payer signer for the Permit2 authorization.
 * @param x402Version - Protocol version for the payment envelope.
 * @param paymentRequirements - Server-provided payment requirements.
 * @param channelConfig - Channel configuration bound into the voucher and witness.
 * @param depositAmount - Token amount deposited into the channel.
 * @param maxClaimableAmount - Cumulative amount signed in the voucher.
 * @param voucherSigner - Optional signer for the voucher.
 * @returns Signed deposit payload and voucher.
 */
export async function createBatchSettlementPermit2DepositPayload(
  signer: ClientEvmSigner,
  x402Version: number,
  paymentRequirements: PaymentRequirements,
  channelConfig: ChannelConfig,
  depositAmount: string,
  maxClaimableAmount: string,
  voucherSigner?: ClientEvmSigner,
): Promise<PaymentPayloadResult> {
  const chainId = getEvmChainId(paymentRequirements.network);
  const nonce = createPermit2Nonce();
  const deadline = Math.floor(Date.now() / 1000 + paymentRequirements.maxTimeoutSeconds).toString();
  const channelId = computeChannelId(channelConfig, paymentRequirements.network);

  const permit2Authorization = {
    from: signer.address,
    permitted: {
      token: getAddress(paymentRequirements.asset),
      amount: depositAmount,
    },
    spender: getAddress(PERMIT2_DEPOSIT_COLLECTOR_ADDRESS),
    nonce,
    deadline,
    witness: {
      channelId,
    },
  };

  const signature = await signer.signTypedData({
    domain: { name: "Permit2", chainId, verifyingContract: PERMIT2_ADDRESS },
    types: batchPermit2WitnessTypes,
    primaryType: "PermitWitnessTransferFrom",
    message: {
      permitted: {
        token: permit2Authorization.permitted.token,
        amount: BigInt(permit2Authorization.permitted.amount),
      },
      spender: permit2Authorization.spender,
      nonce: BigInt(permit2Authorization.nonce),
      deadline: BigInt(permit2Authorization.deadline),
      witness: {
        channelId,
      },
    },
  });

  const voucher = await signVoucher(
    voucherSigner ?? signer,
    channelId,
    maxClaimableAmount,
    paymentRequirements.network,
  );

  const payload: BatchSettlementDepositPayload = {
    type: "deposit",
    channelConfig,
    voucher,
    deposit: {
      amount: depositAmount,
      authorization: {
        permit2Authorization: {
          ...permit2Authorization,
          signature,
        },
      },
    },
  };

  return { x402Version, payload };
}
